### PR TITLE
[DEV-9650] Filter DEFC transactions on public law enactment date

### PR DESCRIPTION
**Pull request by aguest-kc**
**Description:**
Update Advanced Search to use specific dates for each disaster code when filtering transactions.

**Technical details:**
When querying for transactions that are associated with a COVID-19 or IIJA DEFC, only include the transactions that have an `action_date` on/after the disaster code's `earliest_public_law_enactment_date`. Transactions associated with any other DEFC or no DEFC are included as before.

**Requirements for PR merge:**

1. [x] Unit & integration tests updated
2. [ ] API documentation updated
3. [ ] Necessary PR reviewers:
    - [ ] Backend
4. [x] Data validation completed
5. [x] Jira Ticket [DEV-9650](https://federal-spending-transparency.atlassian.net/browse/DEV-9650):
    - [x] Link to this Pull-Request
    - [x] Before / After data comparison

### CODE REVIEW

File: usaspending_api/common/query_with_filters.py [hunk 1]
```diff
@@ -433,30 +435,43 @@ def generate_elasticsearch_query(cls, filter_values: List[str], query_type: _Que
...
+        other_filters = list(set(filter_values) - set(all_covid_iija_defc.keys()))
         other_queries = [ES_Q("match", **{def_code_field: v}) for v in other_filters]
 
-        if covid_queries:
+        covid_iija_filters = {k: all_covid_iija_defc[k] for k in filter_values if k in all_covid_iija_defc.keys()}
+        covid_iija_queries = []
+
+        # Filter on the `disaster_emergency_fund_code` AND `action_date` values for transactions
+        for code, enactment_date in covid_iija_filters.items():
+            if query_type == _QueryType.TRANSACTIONS:
```
Comment (collinwr) [inline usaspending_api/common/query_with_filters.py:453]: Great catch creating separate logic for Transactions when building up the `covid_iija_queries` list

File: usaspending_api/references/management/commands/load_disaster_emergency_fund_codes.py [hunk 2]
```diff
@@ -84,8 +85,24 @@ def _prep(text):
             return text.strip()
         return text
 
-    def _read_raw_def_code_csv(self):
+    @staticmethod
+    def _prep_date(date_string: str) -> Union[datetime, None]:
+        """Convert a string to a datetime object or None if no string was provided.
+        This assumes the date is in the following format: 12-25-2023 (mm-dd-yyyy)
```
Comment (collinwr) [inline usaspending_api/references/management/commands/load_disaster_emergency_fund_codes.py:91]: Looks like in a few places (including `query_with_filters.py` we use the format `yyyy-mm-dd`. Can you please update this (as well as the CSV) to match?

Comment (aguest-kc) [reply usaspending_api/references/management/commands/load_disaster_emergency_fund_codes.py]: Updated the date format

**Review by collinwr — CHANGES_REQUESTED** [review]
PR implementation looks good, but I'd like to see the test cases fleshed out a bit to cover this new functionality before we merge it.

**Review by collinwr — APPROVED** [review]
Latest unit and integration tests look great! Approving

**Pull request merged**